cpp/src/arrow/engine/logical_plan.h

```diff
@@ -0,0 +1,136 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/type_fwd.h"
+#include "arrow/util/compare.h"
+#include "arrow/util/variant.h"
+
+namespace arrow {
+
+namespace dataset {
+class Dataset;
+}
+
+namespace engine {
+
+class Catalog;
+class Expr;
+class ExprType;
+
+class LogicalPlan : public util::EqualityComparable<LogicalPlan> {
+ public:
+  explicit LogicalPlan(std::shared_ptr<Expr> root);
+
+  const std::shared_ptr<Expr>& root() const { return root_; }
+  const ExprType& type() const;
+
+  bool Equals(const LogicalPlan& other) const;
+  std::string ToString() const;
+
+ private:
+  std::shared_ptr<Expr> root_;
+};
+
+struct LogicalPlanBuilderOptions {
+  /// Catalog containing named tables.
+  std::shared_ptr<Catalog> catalog;
+};
+
+class LogicalPlanBuilder {
+ public:
+  using ResultExpr = Result<std::shared_ptr<Expr>>;
+
+  explicit LogicalPlanBuilder(LogicalPlanBuilderOptions options = {});
+
+  /// \defgroup leaf-nodes Leaf nodes in the logical plan
+  /// @{
+
+  /// \brief Construct a Scalar literal.
+  ResultExpr Scalar(const std::shared_ptr<Scalar>& scalar);
+
+  /// \brief References a field by name.
+  ResultExpr Field(const std::shared_ptr<Expr>& input, const std::string& field_name);
+  /// \brief References a field by index.
+  ResultExpr Field(const std::shared_ptr<Expr>& input, int field_index);
+
+  /// \brief Scan a Table/Dataset from the Catalog.
+  ResultExpr Scan(const std::string& table_name);
+
+  /// @}
+
+  /// \defgroup comparator-nodes Comparison operators
+  /// @{
+
+  /*
+  TODO(fsaintjacques): This.
+  ResultExpr Equal(const std::shared_ptr<Expr>& lhs, const std::shared_ptr<Expr>& rhs);
+  ResultExpr NotEqual(const std::shared_ptr<Expr>& lhs, const std::shared_ptr<Expr>& rhs);
+  ResultExpr GreaterThan(const std::shared_ptr<Expr>& lhs,
+                         const std::shared_ptr<Expr>& rhs);
+  ResultExpr GreaterEqualThan(const std::shared_ptr<Expr>& lhs,
+                              const std::shared_ptr<Expr>& rhs);
+  ResultExpr LessThan(const std::shared_ptr<Expr>& lhs, const std::shared_ptr<Expr>& rhs);
+  ResultExpr LessEqualThan(const std::shared_ptr<Expr>& lhs,
+                           const std::shared_ptr<Expr>& rhs);
+  */
+
+  /// @}
+
+  /// \defgroup rel-nodes Relational operator nodes in the logical plan
+
+  /// \brief Filter rows of a relation with the given predicate.
+  ResultExpr Filter(const std::shared_ptr<Expr>& input,
+                    const std::shared_ptr<Expr>& predicate);
+
+  /// \brief Project (mutate) columns with given expressions.
```

Comment:
What is the difference between project and mutate? By mutate, do you mean UPDATE in SQL? (but Arrow data is immutable...)

Reply:
This is an alias for projection. This term is used by dplyr crowd.

Reply:
I thought the same when I read this. Could we leave the dplyr specific alias (`Mutate` method, comment is fine) to R?

Reply:
@fsaintjacques since the other two doccomments use `select` now:
```suggestion
  /// \brief Project (select) columns with given expressions.
```